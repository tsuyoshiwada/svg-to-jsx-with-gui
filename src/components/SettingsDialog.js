import React, { Component } from 'react';
import styled from 'styled-components';
import keycode from 'keycode';
import { easings } from '../styles';
import SettingsForm from './SettingsForm'; // eslint-disable-line import/no-named-as-default
import Button from './internal/Button';
import { Close } from './icons/';

const DialogWrapper = styled.div`
  position: fixed;
  top: 0;
  right: 0;
  bottom: 60px;
  left: 0;
  z-index: 1000;
  padding: 0 30px 120px;
  background: #fff;
  overflow-x: hidden;
  overflow-y: auto;
  transition-property: opacity, visibility, transform;
  transition-timing-function: ${easings.easeOutQuart};

  &:focus {
    outline: none;
  }

  &.dialog-enter {
    opacity: 0;
    visibility: hidden;
    transform: translateY(-15px);
  }

  &.dialog-enter.dialog-enter-active {
    opacity: 1;
    visibility: visible;
    transform: translateY(0);
    transition-duration: 420ms;
  }

  &.dialog-leave {
    opacity: 1;
    visibility: visible;
    transform: translateY(0);
  }

  &.dialog-leave.dialog-leave-active {
    opacity: 0;
    visibility: hidden;
    transform: translateY(-5px);
    transition-duration: 120ms;
  }
`;

const DialogInner = styled.div`
  width: 540px;
  margin: 0 auto;

  & h2 {
    margin: 60px 0;
    font-size: 2.62rem;
  }

  & h3 {
    margin: 60px 0 30px;
    font-size: 1.5rem;

    & small {
      font-size: 0.75rem;
    }
  }

  & code {
    display: inline-block;
    padding: 0.1em 0.4em;
    background: #f0f0f0;
    border-radius: 3px;
    font-size: 0.85em;
    font-family: 'SFMono-Regular', Consolas, 'Liberation Mono', Menlo, Courier, monospace;
  }
`;

const CloseButton = Button.extend`
  position: fixed;
  top: 20px;
  right: 20px;
  width: 70px;
  height: 70px;
  background: #f7f7f7;
  border-radius: 50%;
  box-shadow: none;
  font-size: 0.68rem;
  font-weight: normal;

  & svg {
    display: block;
    width: 14px;
    height: 14px;
    margin: 3px auto 2px;
  }

  &:hover {
    background-color: #fff;
    box-shadow: 0 2px 4px rgba(0, 0, 0, 0.16);
  }

  &:active {
    background-color: #fff;
    box-shadow: 0 1px 2px rgba(0, 0, 0, 0.11);
  }
`;


export default class SettingsDialog extends Component {
  componentDidMount() {
    this.mounted = true;
  }

  componentWillUnmount() {
    this.mounted = false;
  }

  mounted = false;

  handleRef = (dialog) => {
    this.dialog = dialog;
  };

  handleFormRef = (form) => {
    this.form = form;
  };

  handleTransitionEnd = (e) => {
    if (this.mounted && e.target === this.dialog) {
      this.dialog.focus();
    }
  };

  handleKeyDown = (e) => {
    if (keycode(e) === 'esc' && this.form.isValid()) {
      this.props.onRequestClose(this.form.getValues());
    }
  };

  handleCloseClick = (e) => {
    e.preventDefault();

    if (this.form.isValid()) {
      this.props.onRequestClose(this.form.getValues());
    }
  };

  render() {
    const { values, onChange } = this.props;

    return (
      <DialogWrapper
        innerRef={this.handleRef}
        tabIndex="-1"
        onKeyDown={this.handleKeyDown}
        onTransitionEnd={this.handleTransitionEnd}
      >
        <CloseButton onClick={this.handleCloseClick}>
          <div>
            <Close />
            Close
          </div>
        </CloseButton>

        <DialogInner style={this.state}>
          <SettingsForm
            onInitialize={this.handleFormRef}
            values={values}
            onChange={onChange}
          />
        </DialogInner>
      </DialogWrapper>
    );
  }
}
