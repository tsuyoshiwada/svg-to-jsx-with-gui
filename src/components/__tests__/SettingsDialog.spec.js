import React from 'react';
import { shallow } from 'enzyme';
import SettingsDialog, { DialogWrapper, CloseButton } from '../SettingsDialog';


describe('components#<SettingsDialog />', () => {
  test('Should be render', () => {
    const wrapper = shallow(<SettingsDialog />);
    expect(wrapper).toMatchSnapshot();
  });


  test('Should be call onRequestClose when ESC is keyed down', () => {
    const fn = jest.fn();
    const wrapper = shallow(<SettingsDialog onRequestClose={fn} />);

    wrapper.instance().form = {
      isValid: () => true,
      getValues: () => 'foo',
    };

    expect(fn.mock.calls.length).toBe(0);

    wrapper.find(DialogWrapper).simulate('keydown', { keyCode: 27 });

    expect(fn.mock.calls.length).toBe(1);
  });


  test('Should be call onRequestClose when close button click', () => {
    const fn = jest.fn();
    const wrapper = shallow(<SettingsDialog onRequestClose={fn} />);

    wrapper.instance().form = {
      isValid: () => true,
      getValues: () => 'foo',
    };

    expect(fn.mock.calls.length).toBe(0);

    wrapper.find(CloseButton).simulate('click', { preventDefault: () => null });

    expect(fn.mock.calls.length).toBe(1);
  });
});
