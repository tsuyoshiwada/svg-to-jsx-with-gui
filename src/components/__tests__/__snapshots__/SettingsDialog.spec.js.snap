// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components#<SettingsDialog /> Should be render 1`] = `
<styled.div
  innerRef={[Function]}
  onKeyDown={[Function]}
  onTransitionEnd={[Function]}
  tabIndex="-1"
>
  <styled.button
    onClick={[Function]}
  >
    <div>
      <Close />
      Close
    </div>
  </styled.button>
  <styled.div
    style={null}
  >
    <dripForm(SettingsForm)
      onChange={null}
      onClear={null}
      onInitialize={[Function]}
      onInvalidSubmit={null}
      onReset={null}
      onValidSubmit={null}
      values={Object {}}
    />
  </styled.div>
</styled.div>
`;
